transformer_engine/common/transpose/transpose.cu

```diff
@@ -25,7 +27,80 @@ namespace {
 constexpr size_t warps_per_tile = 4;
 constexpr size_t block_size = THREADS_PER_WARP * warps_per_tile;
 
-}  // namespace
+/* Performance heuristics for optimized kernel parameters */
+struct KernelConfig {
+  /** Vector load size */
+  size_t load_size;
+  /** Vector store size */
+  size_t store_size;
+
+  /* Whether config is valid */
+  bool valid = false;
+  /* Number of CUDA blocks */
+  size_t num_blocks = 0;
+
+  /* Number of active SMs */
+  size_t active_sm_count = 0;
+  /* Elements per L1 cache load */
+  size_t elements_per_load = 0;
+  /* Elements per L1 cache store */
+  size_t elements_per_store = 0;
+
+  KernelConfig(size_t row_length,
+               size_t num_rows,
+               size_t type_size,
+               size_t load_size_,
+               size_t store_size_)
+    : load_size{load_size_}
+    , store_size{store_size_} {
+    // Check that tiles are correctly aligned
+    constexpr size_t cache_line_size = 128;
+    if (load_size % type_size != 0
```

Comment:
Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?